specs/governance/gov-token.md

```diff
@@ -24,72 +24,82 @@
 
 `GovernanceToken` is an [ERC20](https://eips.ethereum.org/EIPS/eip-20) token contract that inherits from `ERC20Burnable`,
 `ERC20Votes`, and `Ownable`. It allows token holders to delegate their voting power to other addresses, enabling a representative
-voting system.
+voting system. The contract integrates with the `Alligator` contract through a hook-based approach to enable advanced delegation
+functionality, including partial delegation and subdelegations.
 
-The specific implementation details of the voting and delegation logic are inherited from the `ERC20Votes` contract. The
-`GovernanceToken` contract focuses on integrating this functionality with `ERC20Burnable` and adding the minting
-capability. EIP-712 permit functionality is also inherited from `ERC20Permit`, which is imported by `ERC20Votes`.
+### Hook-based Integration with Alligator
 
-### Checkpoints
+The `GovernanceToken` contract integrates with the `Alligator` contract through a hook-based approach to enable advanced
+delegation features. The `_afterTokenTransfer` function in the `GovernanceToken` is modified to call the `afterTokenTransfer`
+function in the `Alligator` contract, allowing the `Alligator` to consume the hooks and update its delegation and checkpoint
+mappings accordingly.
 
-`ERC20Votes` uses checkpoints to track voting power history. A checkpoint is a "snapshot" of the voting power of an
-address at a specific block number.
+To ensure that token transfers can continue even if the call to the Alligator fails, a try-catch mechanism is implemented
+in the `_afterTokenTransfer` function. If the call to the Alligator's `afterTokenTransfer` function fails, the token transfer
+will still be completed, and an event will be emitted to indicate the failure. This prevents the token transfer from being
+blocked due to issues with the Alligator contract.
 
 ```solidity
-struct Checkpoint {
-    uint32 fromBlock;
-    uint224 votes;
+function _afterTokenTransfer(address from, address to, uint256 amount) internal override {
+    bytes memory data = abi.encodeWithSelector(IAlligator.afterTokenTransfer.selector, from, to, amount);
+    bool success;
+    assembly ("memory-safe") {
+        success := call(gas(), alligatorAddress, 0, add(data, 0x20), mload(data), 0, 0)
+    }
+    
+    if (!success) emit AlligatorCallFailed(from, to, amount);
 }
 ```
 
-Checkpoints are organized as a one-to-one mapping of user addresses to append-only lists of checkpoints. When a
-user is involved in a token transfer, mint, burn, or vote delegation, a new checkpoint is added to their corresponding
-list of checkpoints, reflecting the user's updated voting power at that block number.
+All delegation-related state, including the `delegates`, `checkpoints`, and `numCheckpoints` mappings, is shifted from the
+`GovernanceToken` to the `Alligator` contract. The `Alligator` treats the original checkpoints from the `GovernanceToken`
+as a starting point for its own checkpoints.
```

Comment:
> The `Alligator` treats the original checkpoints from the `GovernanceToken` as a starting point for its own checkpoints.

Can we expand on this? It's not clear to me what this looks like, e.g. what happens to all existing delegation. It sounds like a lot of state needs to be migrated from `GovernanceToken` to `Alligator`?

Reply:
https://github.com/ethereum-optimism/specs/pull/233/commits/8def8fa0af9746ed8f4f6535edcfb9cb98b8668a. lmk if it's clearer!

Reply:
I see, so is this all correct?
1. Alligator always check it's own state first. If not empty, use that. If empty, fetch state from GovernanceToken and set that in its own storage
2. The existing data in GovernanceToken storage will not be cleared and will exist forever
3. Getter methods in GovernanceToken will be changed to not reference GovernanceToken storage vars directly, and instead call alligator to ask for the data. Per (1), Alligator may then call back into GovernanceToken storage
4. We need to make sure there are no public vars exposed in GovernanceToken, as this is a potential way users with existing integrations may continue to read stale data. We should think about any others

Reply:
yes, correct on all points

Reply:
Awesome, thanks. Let's make sure these points are all clear in the spec (maybe it is, I will re-review later)

Reply:
clarified point 3 (and in effect point 1) in 42b51914c0e3dcee164c934986fbd2dfff0e6698

Reply:
Thank you!